frame/ethereum/src/lib.rs

```diff
@@ -86,82 +85,83 @@ impl Get<H256> for IntermediateStateRoot {
+		/// How Ethereum state root is calculated.
+		type StateRoot: Get<H256>;
 	}
-}
 
-decl_event!(
-	/// Ethereum pallet events.
-	pub enum Event {
+	/// Current building block's transactions and receipts.
+	#[pallet::storage]
+	#[pallet::getter(fn pending)]
+	pub type Pending<T: Config> = StorageValue<_, Vec<(ethereum::Transaction, TransactionStatus, ethereum::Receipt)>, ValueQuery>;
+
+	/// The current Ethereum block.
+	#[pallet::storage]
+	#[pallet::getter(fn current_block)]
+	pub type CurrentBlock<T: Config> = StorageValue<_, Option<ethereum::Block>, ValueQuery>;
+
+	/// The current Ethereum receipts.
+	#[pallet::storage]
+	#[pallet::getter(fn current_receipts)]
+	pub type CurrentReceipts<T: Config> = StorageValue<_, Option<Vec<ethereum::Receipt>>, ValueQuery>;
+
+	/// The current transaction statuses.
+	#[pallet::storage]
+	#[pallet::getter(fn current_transaction_statuses)]
+	pub type CurrentTransactionStatuses<T: Config> = StorageValue<_, Option<Vec<TransactionStatus>>, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
 		/// An ethereum transaction was successfully executed. [from, to/contract_address, transaction_hash, exit_reason]
 		Executed(H160, H160, H256, ExitReason),
 	}
-);
 
-
-decl_error! {
-	/// Ethereum pallet errors.
-	pub enum Error for Module<T: Config> {
+	#[pallet::error]
+	pub enum Error<T> {
 		/// Signature is invalid.
 		InvalidSignature,
 		/// Pre-log is present, therefore transact is not allowed.
 		PreLogExists,
 	}
-}
 
-decl_module! {
-	/// Ethereum pallet module.
-	pub struct Module<T: Config> for enum Call where origin: T::Origin {
-		/// Deposit one of this pallet's events by using the default implementation.
-		fn deposit_event() = default;
-
-		/// Transact an Ethereum transaction.
-		#[weight = <T as pallet_evm::Config>::GasWeightMapping::gas_to_weight(transaction.gas_limit.unique_saturated_into())]
-		fn transact(origin, transaction: ethereum::Transaction) -> DispatchResultWithPostInfo {
-			ensure_none(origin)?;
-
-			Self::do_transact(transaction)
-		}
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
 
-		fn on_finalize(n: T::BlockNumber) {
-			<Module<T>>::store_block(
-				fp_consensus::find_pre_log(&frame_system::Module::<T>::digest()).is_err(),
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_finalize(_n: T::BlockNumber) {
+			<Pallet<T>>::store_block(
+				fp_consensus::find_pre_log(&frame_system::Pallet::<T>::digest()).is_err(),
 			);
 		}
 
-		fn on_initialize(n: T::BlockNumber) -> Weight {
-			Pending::kill();
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			// Calculate the ethereum genesis block
+			<Pallet<T>>::store_block(false);
+
+			// Initialize the storage schema at the well known key.
+			frame_support::storage::unhashed::put::<EthereumStorageSchema>(&PALLET_ETHEREUM_SCHEMA, &EthereumStorageSchema::V1);
```

Comment:
There is some difference from the original implementation, any reason?

Reply:
Looks like it was updated, I'll update too.

Reply:
Updated